File: 46. Permutations.py
# https://leetcode.com/problems/permutations/

# Given an array nums of distinct integers, return all the possible
# permutations. You can return the answer in any order.

###############################################################################

# all combos -> dfs -> add child that is not used

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        if not nums or len(nums) == 0: return []
        
        ans = []
        is_used = set()
        self.dfs(ans, [], nums, is_used)
        return ans
        
    def dfs(self, ans, holder, nums, is_used):
        if len(holder) == len(nums):
            ans.append(holder.copy())
            return
        
        # generate child
        for num in nums:
            if num not in is_used: # no reuse
                holder.append(num)
                is_used.add(num)
                self.dfs(ans, holder, nums, is_used)
                holder.pop()
                is_used.remove(num)

File: 46. Permutations.cpp
// https://leetcode.com/problems/permutations/

// Given an array nums of distinct integers, return all the possible permutations.
// You can return the answer in any order.

////////////////////////////////////////////////////////////////////////////////

// dfs
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        n = nums.size();
        visited = vector<bool>(n, false);
        dfs(nums);
        return ans;
    }
private:
    int n;
    vector<bool> visited;
    vector<int> holder;
    vector<vector<int>> ans;
    void dfs(const vector<int>& nums) {
        if (holder.size() == n) {
            ans.push_back(holder); return;
        }
        for (int i = 0; i < n; ++i) {
            if (!visited[i]) {
                visited[i] = true;
                holder.push_back(nums[i]);
                dfs(nums);
                holder.pop_back();
                visited[i] = false;
            }
        }
        return;
    }
};
